backend/composer/services/cs_ingestion/cs_ingestion_services.py

```diff
@@ -1,292 +1,522 @@
-
-            for dest in statement[DESTINATION]:
-                destination_entity = AnatomicalEntity.objects.filter(ontology_uri=dest[ENTITY_URI]).first()
-                destination_type = dest.get(TYPE, DestinationType.UNKNOWN)
-
-                if destination_entity:
-                    destination_instance, _ = Destination.objects.get_or_create(
-                        connectivity_statement=connectivity_statement,
-                        defaults={
-                            "type": destination_type
-                        }
-                    )
-                    destination_instance.anatomical_entities.add(destination_entity)
-
-            vias_data = [
-                Via(connectivity_statement=connectivity_statement, type=via[TYPE], order=index)
-                for index, via in enumerate(statement[VIAS])
-            ]
-
-            created_vias = Via.objects.bulk_create(vias_data)
-
-            for via_instance, via_data in zip(created_vias, statement[VIAS]):
-                anatomical_entities = AnatomicalEntity.objects.filter(
-                    ontology_uri__in=via_data[ENTITY_URI]
-                )
-                via_instance.anatomical_entities.set(anatomical_entities)
-
-            provenances_list = statement[PROVENANCE][0].split(", ") if has_prop(statement[PROVENANCE]) else [
-                statement[ID]]
-            provenances = (Provenance(connectivity_statement=connectivity_statement, uri=provenance) for provenance in
-                           provenances_list)
-            Provenance.objects.bulk_create(provenances)
-            # only 5 statements have a note_alert but they were all filtered out since at least one of their anatomical entities was not found in composer db
-            if has_prop(statement[NOTE_ALERT]):
-                Note.objects.create(connectivity_statement=connectivity_statement,
-                                    user=User.objects.get(username="system"), type=NoteType.ALERT,
-                                    note=statement[NOTE_ALERT][0])
-            connectivity_statement.save()
-
-            # transitions sentence from  open --> compose_now, and statement from draft --> compose_now --> exported 
-            do_state_transitions(sentence)
+
+        for backward_cs in backward_connections:
+            current_reason = ''
+            if backward_cs.state != CSState.INVALID:
+                # Build the reason string
+                current_reason = (f"statement with id {backward_cs.id} is invalid because its "
+                                  f"forward connection with id {connectivity_statement.id} is invalid")
+                if previous_reason:
+                    current_reason += f" because {previous_reason}"
+
+                # Transition the backward statement to an invalid state with the built reason
+                do_transition_to_invalid(backward_cs, current_reason)
+            else:
+                # TODO: Should we update the reason?
```

Comment:
If a statement is already invalid per se and should also be marked as invalid due to a invalid forward connection, should we just have 1 note for the per se or add a new note for the forward connection as well?

Reply:
@afonsobspinto as said during the last meeting I would split the ingestion into steps. For example 
step 1 is load and transform all KS from neurondm into "an internal format" in memory
step 2 load all KS (and set the correct state of these KS) to the database
step 3 post ingestion step to update KS into (in)validate depending on depending KS  validations

I suggest that step 3 has one note per ingestion cyclus per KS with the reason that it's invalidated because of "the list of all invalid dependencies (path)"

cc: @ddelpiano

Reply:
The current implementation is already following that flow. The question I have here is, in step 2 we will have connectivity statements set to the Invalid state and we will add a note to it saying why.
Then, on step 3, which is optional and configurable by a flag `update_upstream`, it can happen that one of the connectivity statements we set to Invalid in 2., also should become Invalid because of dependencies. In this case, do we add a new note?

Reply:
imo we should add another note yes